pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ThresholdShedder.java

```diff
@@ -233,5 +237,23 @@ private Pair<Boolean, String> getMaxUsageBroker(
         }
         return Pair.of(hasBrokerBelowLowerBound, maxUsageBrokerName);
     }
+    @Override
+    public void onBrokerChange(Set<String> newBrokers) {
+        synchronized (activeBrokers) {
+            activeBrokers.clear();
+            activeBrokers.addAll(newBrokers);
+        }
+    }
+
+    private void cleanUnActiveBroker() {
+        if (!activeBrokers.isEmpty()) {
+            synchronized (activeBrokers) {
+                if (!activeBrokers.isEmpty()) {
+                    brokerAvgResourceUsage.keySet().removeIf((key) -> !activeBrokers.contains(key));
```

Comment:
Can we clean this map in the `onBrokerChange` method?

Reply:
Also, we can use `keySet().retainAll(activeBrokers)` to get Intersection.

Reply:
> Can we clean this map in the onBrokerChange method?

If we clean the map in `onBrokerChange`，we have to lock every time, but now we don't need to lock every time

> Also, we can use keySet().retainAll(activeBrokers) to get Intersection.

We don't want to clean up the inactive Broker every time, because it needs to be locked. 
If we don't clear this collection, we have to lock it every time to judge whether it needs to be cleaned up

Reply:
I see. Can we change the `brokerAvgResourceUsage` to `ConcurrentHashMap`, and remove the inactive broker in the `onBrokerChange` method?

```
@Override
    public void onBrokerChange(Set<String> newBrokers) {
          brokerAvgResourceUsage.keySet().retainAll(newBrokers);
    }

```

Reply:
In this way, the place where `brokerAvgResourceUsage` is used in `findBundlesForUnloading` must be locked, that is, it must be locked every time, because now `brokerAvgResourceUsage` will be called in the `updateAvgResourceUsage` method every time.

Reply:
Ok, in my opinion, `findBundlesForUnloading` method only calls in the leader on one thread. The default Load shedding interval is 1 minute, so it should not be a hot path.

But the current solution is also ok with me.